### PR TITLE
Saliency

### PR DESCRIPTION
**Pull request by corochann**
saliency modules

[Calculator]
 - gradient
 - integrated gradient
 - occlusion

[Visualizer]
 - mol, smiles
 - image
 - table data

[Addtional utils]
 - VariableMonitorLinkHook
 - TrainConfigFunctionHook

### CODE REVIEW

File: chainer_chemistry/saliency/visualizer/mol_visualizer.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        Chem.SanitizeMol(mol)
+        Chem.Kekulize(mol)
+        num_atoms = mol.GetNumAtoms()
+
+        # --- type check ---
+        if not saliency.ndim == 1:
+            raise ValueError("Unexpected value saliency.shape={}"
+                             .format(saliency.shape))
+
+        # Cut saliency array for unnecessary tail part
```
[the remark this answers is not in the document]
Comment (corochann) [reply chainer_chemistry/saliency/visualizer/mol_visualizer.py]: it is expected. atom_array is often 0 padded with concat_mols, so we need to truncate this padded length.

File: chainer_chemistry/saliency/visualizer/mol_visualizer.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+                scaled `x`, for plotting.
+            legend (str): legend for the plot
+        """
+        rdDepictor.Compute2DCoords(mol)
+        Chem.SanitizeMol(mol)
+        Chem.Kekulize(mol)
+        num_atoms = mol.GetNumAtoms()
+
+        # --- type check ---
+        if not saliency.ndim == 1:
```
[the remark this answers is not in the document]
Comment (corochann) [reply chainer_chemistry/saliency/visualizer/mol_visualizer.py]: done

File: chainer_chemistry/saliency/visualizer/common.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    xp = cuda.get_array_module(saliency)
+    maxv = xp.max(xp.abs(saliency))
+    if maxv <= 0:
+        return numpy.zeros_like(saliency)
+    else:
+        return saliency / maxv
+
+
+def normalize_scaler(saliency, axis=None):
+    """Normalize saliency to be sum=1
```
[the remark this answers is not in the document]
Comment (corochann) [reply chainer_chemistry/saliency/visualizer/common.py]: modified to show warning message

File: chainer_chemistry/saliency/calculator/base_calculator.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+    link_hooks[name] = linkhook
+    linkhook.added(None)
+    return linkhook
+
+
+def delete_linkhook(linkhook, prefix=''):
+    name = prefix + linkhook.name
+    link_hooks = chainer._get_link_hooks()
+    if name not in link_hooks.keys():
+        print('[WARNING] linkhook {} is not registered'.format(name))
```
[the remark this answers is not in the document]
Comment (corochann) [reply chainer_chemistry/saliency/calculator/base_calculator.py]: done

File: chainer_chemistry/saliency/calculator/base_calculator.py [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+
+        elem_list = [elem for batch in batch_list for elem in batch]
+        return _concat_arrays_with_padding(elem_list, padding=0)
+
+
+def add_linkhook(linkhook, prefix=''):
+    link_hooks = chainer._get_link_hooks()
+    name = prefix + linkhook.name
+    if name in link_hooks:
+        print('[WARNING] hook {} already exists, overwrite.'.format(name))
```
[the remark this answers is not in the document]
Comment (corochann) [reply chainer_chemistry/saliency/calculator/base_calculator.py]: done

File: tests/saliency_tests/visualizer_tests/test_image_visualizer.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    visualizer = ImageVisualizer()
+
+    # 1. test with setting save_filepath
+    save_filepath = os.path.join(str(tmpdir), 'tmp.png')
+    visualizer.visualize(saliency, save_filepath=save_filepath)
+    assert os.path.exists(save_filepath)
+
+    # 2. test with `save_filepath=None` runs without error
+    image = numpy.random.uniform(0, 1, (ch, h, w))
+    visualizer.visualize(
```
Comment (mottodora) [inline tests/saliency_tests/visualizer_tests/test_image_visualizer.py]: Visualilze method stops unit testing.
Can you insert `plt.ion()` and `plt.close()`? like this.
```
    plt.ion()
    visualizer.visualize(
        saliency, save_filepath=None, feature_names=['hoge', 'huga', 'piyo'],
        num_visualize=2)
    plt.close()
```

File: tests/saliency_tests/visualizer_tests/test_table_visualizer.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    # Only test file is saved without error
+    saliency = numpy.array([0.5, 0.3, 0.2])
+    visualizer = TableVisualizer()
+
+    # 1. test with setting save_filepath
+    save_filepath = os.path.join(str(tmpdir), 'tmp.png')
+    visualizer.visualize(saliency, save_filepath=save_filepath)
+    assert os.path.exists(save_filepath)
+    # 2. test with `save_filepath=None` runs without error
+    visualizer.visualize(
```
Comment (mottodora) [inline tests/saliency_tests/visualizer_tests/test_table_visualizer.py]: As same as image_visualizer

**Comment by corochann** [comment]
Updated based on comment!

**Pull request merged**